File: src/logic.js
// logic.js
// Converted from logic.py

const MINOR_HEARTS = Array.from({length: 6}, (_,i) => `${2+i}H`);   // 2..7
const MINOR_DIAMONDS = Array.from({length: 6}, (_,i) => `${2+i}D`);
const MINOR_SPADES = Array.from({length: 6}, (_,i) => `${2+i}S`);
const MINOR_CLUBS = Array.from({length: 6}, (_,i) => `${2+i}C`);

const MAJOR_HEARTS = [...Array.from({length: 2}, (_,i) => `${9+i}H`), "JH","QH","KH","AH"];
const MAJOR_DIAMONDS = [...Array.from({length: 2}, (_,i) => `${9+i}D`), "JD","QD","KD","AD"];
const MAJOR_SPADES = [...Array.from({length: 2}, (_,i) => `${9+i}S`), "JS","QS","KS","AS"];
const MAJOR_CLUBS = [...Array.from({length: 2}, (_,i) => `${9+i}C`), "JC","QC","KC","AC"];

const EIGHTS = ["8H","8D","8S","8C","JJ","JG"];

export const ALL_CARDS = [
  ...MINOR_HEARTS, ...MINOR_DIAMONDS, ...MINOR_SPADES, ...MINOR_CLUBS,
  ...MAJOR_HEARTS, ...MAJOR_DIAMONDS, ...MAJOR_SPADES, ...MAJOR_CLUBS,
  ...EIGHTS
];

// findPit: returns { pitName, pitCards } or { pitName: "UNKNOWN" }
export function findPit(card) {
  if (MINOR_HEARTS.includes(card)) return { pitName: "MINOR_HEARTS", pitCards: MINOR_HEARTS };
  if (MINOR_DIAMONDS.includes(card)) return { pitName: "MINOR_DIAMONDS", pitCards: MINOR_DIAMONDS };
  if (MINOR_SPADES.includes(card)) return { pitName: "MINOR_SPADES", pitCards: MINOR_SPADES };
  if (MINOR_CLUBS.includes(card)) return { pitName: "MINOR_CLUBS", pitCards: MINOR_CLUBS };
  if (MAJOR_HEARTS.includes(card)) return { pitName: "MAJOR_HEARTS", pitCards: MAJOR_HEARTS };
  if (MAJOR_DIAMONDS.includes(card)) return { pitName: "MAJOR_DIAMONDS", pitCards: MAJOR_DIAMONDS };
  if (MAJOR_SPADES.includes(card)) return { pitName: "MAJOR_SPADES", pitCards: MAJOR_SPADES };
  if (MAJOR_CLUBS.includes(card)) return { pitName: "MAJOR_CLUBS", pitCards: MAJOR_CLUBS };
  if (EIGHTS.includes(card)) return { pitName: "EIGHTS", pitCards: EIGHTS };
  return { pitName: "UNKNOWN" };
}

// findTeam: returns team key (e.g. "team_1") or null
export function findTeam(player, teams) {
  for (const [teamKey, members] of Object.entries(teams)) {
    // members includes player-keys and a "total_pits" field
    if (player in members && player !== "total_pits") return teamKey;
  }
  return null;
}

export default class GameState {
  constructor() {
    this.players = {};      // { playerName: [cards] }
    this.teams = {};        // { team_1: { playerName: {...}, total_pits: number }, team_2: {...} }
    this.caller = null;
    this.droppedPits = [];  // list of pit names
    this.logs = [];         // history array of small objects like {CALL: "..." }
  }

  // Set player names. (Defaults to null if not provided, like original.)
  setPlayers(player_1A = null, player_1B = null, player_1C = null, player_2A = null, player_2B = null, player_2C = null) {
    this.players = {
      [player_1A]: [], [player_1B]: [], [player_1C]: [],
      [player_2A]: [], [player_2B]: [], [player_2C]: []
    };

    this.teams = {
      team_1: {
        [player_1A]: { pits_dropped: [], pits_burned: [] },
        [player_1B]: { pits_dropped: [], pits_burned: [] },
        [player_1C]: { pits_dropped: [], pits_burned: [] },
        total_pits: 0
      },
      team_2: {
        [player_2A]: { pits_dropped: [], pits_burned: [] },
        [player_2B]: { pits_dropped: [], pits_burned: [] },
        [player_2C]: { pits_dropped: [], pits_burned: [] },
        total_pits: 0
      }
    };
  }

  // Fisher-Yates shuffle, deal 9 cards to each player
  setShuffleCards() {
    const deck = ALL_CARDS.slice();
    for (let i = deck.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [deck[i], deck[j]] = [deck[j], deck[i]];
    }

    const playerNames = Object.keys(this.players);
    const cardsPerPlayer = 9;

    for (let i = 0; i < playerNames.length; i++) {
      const player = playerNames[i];
      const start = i * cardsPerPlayer;
      const end = start + cardsPerPlayer;
      this.players[player] = deck.slice(start, end);
    }

    console.info("Cards shuffled and dealt to players.");
    this.logs.push({ INITIAL: "Shuffled cards to players." });
  }

  setCaller(player) {
    if (!Object.prototype.hasOwnProperty.call(this.players, player)) {
      throw new Error("Player not found in the game state.");
    }
    this.caller = player;
    console.info(`Caller set to ${player}.`);
    this.logs.push({ CALLER_SET: `Caller set to ${player}.` });
  }

  makeCall(caller, callee, card) {
    if (caller !== this.caller) throw new Error("Check caller again!");
    if (!Object.prototype.hasOwnProperty.call(this.players, caller) || !Object.prototype.hasOwnProperty.call(this.players, callee)) {
      throw new Error("Caller or callee not found in the game state.");
    }

    const callerTeam = findTeam(caller, this.teams);
    const calleeTeam = findTeam(callee, this.teams);
    if (callerTeam === calleeTeam) throw new Error("Calls can only be made to players on the other team.");

    if (!ALL_CARDS.includes(card)) throw new Error("Check card again!");

    const { pitName, pitCards } = findPit(card);
    if (pitName === "UNKNOWN") throw new Error("Card not found in any pit.");

    // Caller called a card they have: pit burn
    if (this.players[caller].includes(card)) {
      console.warn(`Possible pit burn: ${pitName}.`);
      console.warn("Reason: Caller called a card they have.");
    }

    // Caller has no card from that pit: pit burn.
    // (Fixed: check intersection with pitCards, original python had a bug using set(pit).)
    const callerHasAnyFromPit = pitCards.some(c => this.players[caller].includes(c));
    if (!callerHasAnyFromPit) {
      console.warn(`Possible pit burn: ${pitName}`);
      console.warn("Reason: Caller has no card from the pit called");
    }

    // If callee has the card, transfer it to the caller.
    if (this.players[callee].includes(card)) {
      // remove card from callee
      this.players[callee] = this.players[callee].filter(c => c !== card);
      // add to caller
      this.players[caller].push(card);
      console.info(`Card ${card} transferred from ${callee} to ${caller}.`);
      this.logs.push({ CALL: `${caller} called ${callee} for ${card}.` });
    } else {
      // shift caller
      this.caller = callee;
      console.info(`Caller shifted to ${callee} as they do not have the card ${card}.`);
      this.logs.push({ CALL: `${caller} called ${callee} for ${card}.` });
    }
  }

  setPitBurn(burner, card) {
    const { pitName, pitCards } = findPit(card);
    if (pitName === "UNKNOWN") throw new Error("Card not found in any pit.");

    if (!Object.prototype.hasOwnProperty.call(this.players, burner)) {
      throw new Error("Burner not found in the game state.");
    }

    // Remove all cards from all players' hands that belong to the pit.
    for (const p of Object.keys(this.players)) {
      this.players[p] = this.players[p].filter(c => !pitCards.includes(c));
    }

    // Add the pit to droppedPits and update team records
    if (!this.droppedPits.includes(pitName)) {
      this.droppedPits.push(pitName);

      const team = findTeam(burner, this.teams);
      if (team === null) throw new Error("Burner not found in any team.");

      // ensure player's record exists
      if (!this.teams[team][burner]) this.teams[team][burner] = { pits_dropped: [], pits_burned: [] };
      this.teams[team][burner].pits_burned.push(pitName);

      const opposing = team === "team_1" ? "team_2" : "team_1";
      this.teams[opposing].total_pits += 1;
    }

    console.info(`Pit ${pitName} burned by ${burner}.`);
    this.logs.push({ BURN: `${burner} burned pit ${pitName}.` });
  }

  setPitDrop(dropper, card) {
    const { pitName, pitCards } = findPit(card);
    if (pitName === "UNKNOWN") {
      console.error("Card not found in any pit.");
      throw new Error("Card not found in any pit.");
    }

    if (!Object.prototype.hasOwnProperty.call(this.players, dropper)) {
      console.error("Dropper not found in the game state.");
      throw new Error("Dropper not found in the game state.");
    }

    if (this.droppedPits.includes(pitName)) {
      console.warn(`Pit ${pitName} has already been dropped.`);
      return;
    }

    // Remove all cards from all players' hands that belong to the pit.
    for (const p of Object.keys(this.players)) {
      this.players[p] = this.players[p].filter(c => !pitCards.includes(c));
    }

    if (!this.droppedPits.includes(pitName)) {
      this.droppedPits.push(pitName);

      const team = findTeam(dropper, this.teams);
      if (team === null) throw new Error("Dropper not found in any team.");

      if (!this.teams[team][dropper]) this.teams[team][dropper] = { pits_dropped: [], pits_burned: [] };
      this.teams[team][dropper].pits_dropped.push(pitName);

      this.teams[team].total_pits += 1;
    }

    console.info(`Pit ${pitName} dropped by ${dropper}.`);
    this.logs.push({ DROP: `${dropper} dropped pit ${pitName}.` });
  }

  passCard(passer, passee) {
    if (passer !== this.caller) {
      console.error("Passer must be the current caller.");
      throw new Error("Passer must be the current caller.");
    }

    if (!Object.prototype.hasOwnProperty.call(this.players, passer) || !Object.prototype.hasOwnProperty.call(this.players, passee)) {
      console.error("Passer or passee not found in the game state.");
      throw new Error("Passer or passee not found in the game state.");
    }

    const passerTeam = findTeam(passer, this.teams);
    const passeeTeam = findTeam(passee, this.teams);
    if (passerTeam !== passeeTeam) {
      console.error("Cards can only be passed to players on the same team.");
      throw new Error("Cards can only be passed to players on the same team.");
    }

    if (this.players[passer].length !== 1) {
      console.error("Passer must have exactly one card to pass.");
      throw new Error("Passer must have exactly one card to pass.");
    }

    const card = this.players[passer].pop();
    this.players[passee].push(card);

    console.info(`Card ${card} passed from ${passer} to ${passee}.`);
    this.caller = passee;
    console.info(`Caller shifted to ${passee}.`);
    this.logs.push({ PASS: `${passer} passed card ${card} to ${passee}.` });
  }

  forceCardShift(fromPlayer, toPlayer, card) {
    if (!Object.prototype.hasOwnProperty.call(this.players, fromPlayer) || !Object.prototype.hasOwnProperty.call(this.players, toPlayer)) {
      console.error("Both players must be in the game.");
      throw new Error("Both players must be in the game.");
    }

    if (!this.players[fromPlayer].includes(card)) {
      console.error(`Card ${card} not found in ${fromPlayer}'s hand.`);
      throw new Error(`Card ${card} not found in ${fromPlayer}'s hand.`);
    }

    // Transfer the card
    this.players[toPlayer].push(card);
    this.players[fromPlayer] = this.players[fromPlayer].filter(c => c !== card);

    console.info(`Card ${card} transferred from ${fromPlayer} to ${toPlayer}.`);
    this.logs.push({ FORCE_SHIFT: `Clean-up; shifted ${card} from ${fromPlayer} to ${toPlayer}.` });
  }

  resetGameState() {
    this.players = {};
    this.teams = {};
    this.caller = null;
    this.droppedPits = [];
    this.logs = [];
    console.info("Game state has been reset.");
  }
}

File: src/logic.py
import random
import logging

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

MINOR_HEARTS = [f"{i}" + "H" for i in range(2, 7+1)]
MINOR_DIAMONDS = [f"{i}" + "D" for i in range(2, 7+1)]
MINOR_SPADES = [f"{i}" + "S" for i in range(2, 7+1)]
MINOR_CLUBS = [f"{i}" + "C" for i in range(2, 7+1)]
MAJOR_HEARTS = [f"{i}" + "H" for i in range(9, 10+1)] + ["JH", "QH", "KH", "AH"]
MAJOR_DIAMONDS = [f"{i}" + "D" for i in range(9, 10+1)] + ["JD", "QD", "KD", "AD"]
MAJOR_SPADES = [f"{i}" + "S" for i in range(9, 10+1)] + ["JS", "QS", "KS", "AS"]
MAJOR_CLUBS = [f"{i}" + "C" for i in range(9, 10+1)] + ["JC", "QC", "KC", "AC"]
EIGHTS = ["8H", "8D", "8S", "8C", "JJ", "JG"]
ALL_CARDS = MINOR_HEARTS + MINOR_DIAMONDS + MINOR_SPADES + MINOR_CLUBS + MAJOR_HEARTS + MAJOR_DIAMONDS + MAJOR_SPADES + MAJOR_CLUBS + EIGHTS


# Helper function to find the pit of a card.
def find_pit(card):
    if card in MINOR_HEARTS:
        return "MINOR_HEARTS", MINOR_HEARTS
    elif card in MINOR_DIAMONDS:
        return "MINOR_DIAMONDS", MINOR_DIAMONDS
    elif card in MINOR_SPADES:
        return "MINOR_SPADES", MINOR_SPADES
    elif card in MINOR_CLUBS:
        return "MINOR_CLUBS", MINOR_CLUBS
    elif card in MAJOR_HEARTS:
        return "MAJOR_HEARTS", MAJOR_HEARTS
    elif card in MAJOR_DIAMONDS:
        return "MAJOR_DIAMONDS", MAJOR_DIAMONDS
    elif card in MAJOR_SPADES:
        return "MAJOR_SPADES", MAJOR_SPADES
    elif card in MAJOR_CLUBS:
        return "MAJOR_CLUBS", MAJOR_CLUBS
    elif card in EIGHTS:
        return "EIGHTS", EIGHTS
    else:
        return "UNKNOWN"
    
def find_team(player, teams):
    for team, members in teams.items():
        if player in members:
            return team
    return None


class GameState:
    def __init__(self):
        self.players = {}
        self.teams = {}
        self.caller = None
        self.dropped_pits = []
        self.logs = []

    # Set the player names in the game.
    def set_players(self, player_1A = None, player_1B = None, player_1C = None, player_2A = None, player_2B = None, player_2C = None):
        self.players = {
            player_1A : [],
            player_1B : [],
            player_1C : [],
            player_2A : [],
            player_2B : [],
            player_2C : []
        }
        self.teams["team_1"] = {
            player_1A : {"pits_dropped" : [], "pits_burned" : []},
            player_1B : {"pits_dropped" : [], "pits_burned" : []},
            player_1C : {"pits_dropped" : [], "pits_burned" : []},
            "total_pits" : 0
        }
        self.teams["team_2"] = {
            player_2A : {"pits_dropped" : [], "pits_burned" : []},
            player_2B : {"pits_dropped" : [], "pits_burned" : []},
            player_2C : {"pits_dropped" : [], "pits_burned" : []},
            "total_pits" : 0
        }

    # Distribute 9 cards to each player.
    def set_shuffle_cards(self):
        deck = ALL_CARDS.copy()
        random.shuffle(deck)

        player_names = list(self.players.keys())
        cards_per_player = 9

        for i, player in enumerate(player_names):
            start = i * cards_per_player
            end = start + cards_per_player
            self.players[player] = deck[start:end]

        logging.info("Cards shuffled and dealt to players.")

        self.logs.append({"INITIAL":"Shuffled cards to players."})

    # Set the current caller. Use only at start of game.
    def set_caller(self, player):
        if player not in self.players:
            raise ValueError("Player not found in the game state.")
        self.caller = player
        logging.info(f"Caller set to {player}.")
        self.logs.append({"CALLER_SET":f"Caller set to {player}."})

    # Log a call.
    def make_call(self, caller, callee, card):
        # Caller has to make the call.
        if caller != self.caller:
            raise ValueError("Check caller again!")
        
        # Players must exist.
        if caller not in self.players or callee not in self.players:
            raise ValueError("Caller or callee not found in the game state.")
        
        # Ensure calls can only be made to players on the other team.
        caller_team = find_team(caller, self.teams)
        callee_team = find_team(callee, self.teams)
        if caller_team == callee_team:
            raise ValueError("Calls can only be made to players on the other team.")

        if card not in ALL_CARDS:
            raise ValueError("Check card again!")
        
        pit, _ = find_pit(card)
        if pit == "UNKNOWN":
            raise ValueError("Card not found in any pit.")
        
        # Caller called a card they have: pit burn.
        if card in self.players[caller]:
            logging.warning(f"Possible pit burn: {pit}.")
            logging.warning("Reason: Caller called a card they have.")

        # Caller has no card from that pit: pit burn.
        if set(pit).isdisjoint(self.players[caller]):
            logging.warning(f"Possible pit burn: {pit}")
            logging.warning("Reason: Caller has no card from the pit called")

        # If callee has the card, transfer it to the caller.
        if card in self.players[callee]:
            self.players[callee].remove(card)
            self.players[caller].append(card)
            logging.info(f"Card {card} transferred from {callee} to {caller}.")
            self.logs.append({"CALL":f"{caller} called {callee} for {card}."})

        else:
            self.caller = callee
            logging.info(f"Caller shifted to {callee} as they do not have the card {card}.")
            self.logs.append({"CALL":f"{caller} called {callee} for {card}."})


    # Set a pit burn.
    def set_pit_burn(self, burner, card):
        pit, pit_cards = find_pit(card)
        if pit == "UNKNOWN":
            raise ValueError("Card not found in any pit.")
        
        # Player must exist.
        if burner not in self.players:
            raise ValueError("Burner not found in the game state.")

        # Remove all cards from all players' hands that belong to the pit.
        for player in self.players:
            self.players[player] = [c for c in self.players[player] if c not in pit_cards]

        # Add the pit to the dropped pits list.
        if pit not in self.dropped_pits:
            self.dropped_pits.append(pit)
            # Append the burned pit to the player's record in self.teams.
            team = find_team(burner, self.teams)
            if team is None:
                raise ValueError("Burner not found in any team.")

            self.teams[team][burner]["pits_burned"].append(pit)

            # Add a pits_dropped point for the opposing team.
            opposing_team = "team_1" if team == "team_2" else "team_2"
            self.teams[opposing_team]["total_pits"] += 1

        logging.info(f"Pit {pit} burned by {burner}.")
        self.logs.append({"BURN":f"{burner} burned pit {pit}."})


    # Set a pit drop.
    def set_pit_drop(self, dropper, card):
        pit, pit_cards = find_pit(card)
        if pit == "UNKNOWN":
            logging.error("Card not found in any pit.")
            raise ValueError("Card not found in any pit.")
        
        if dropper not in self.players:
            logging.error("Dropper not found in the game state.")
            raise ValueError("Dropper not found in the game state.")

        if pit in self.dropped_pits:
            logging.warning(f"Pit {pit} has already been dropped.")
            return

        # Remove all cards from all players' hands that belong to the pit.
        for player in self.players:
            self.players[player] = [c for c in self.players[player] if c not in pit_cards]

        # Add the pit to the dropped pits list.
        if pit not in self.dropped_pits:
            self.dropped_pits.append(pit)
            # Append the dropped pit to the player's record in self.teams.
            team = find_team(dropper, self.teams)
            if team is None:
                raise ValueError("Dropper not found in any team.")

            self.teams[team][dropper]["pits_dropped"].append(pit)

            # Increment the team's total pits count.
            self.teams[team]["total_pits"] += 1

        logging.info(f"Pit {pit} dropped by {dropper}.")
        self.logs.append({"DROP":f"{dropper} dropped pit {pit}."})

    
    def pass_card(self, passer, passee):
        # Passer must be the current caller.
        if passer != self.caller:
            logging.error("Passer must be the current caller.")
            raise ValueError("Passer must be the current caller.")

        # Both passer and passee must exist in the game.
        if passer not in self.players or passee not in self.players:
            logging.error("Passer or passee not found in the game state.")
            raise ValueError("Passer or passee not found in the game state.")
        
        # Ensure cards can only be passed to players on the same team.
        passer_team = find_team(passer, self.teams)
        passee_team = find_team(passee, self.teams)
        if passer_team != passee_team:
            logging.error("Cards can only be passed to players on the same team.")
            raise ValueError("Cards can only be passed to players on the same team.")

        # Passer must have exactly one card to pass.
        if len(self.players[passer]) != 1:
            logging.error("Passer must have exactly one card to pass.")
            raise ValueError("Passer must have exactly one card to pass.")

        # Transfer the card from passer to passee.
        card = self.players[passer].pop()
        self.players[passee].append(card)
        logging.info(f"Card {card} passed from {passer} to {passee}.")

        # Set the passee as the new caller.
        self.caller = passee
        logging.info(f"Caller shifted to {passee}.")
        self.logs.append({"PASS":f"{passer} passed card {card} to {passee}."})


    # Use if mismatch of cards.
    def force_card_shift(self, from_player, to_player, card):
        # Both players must exist in the game.
        if from_player not in self.players or to_player not in self.players:
            logging.error("Both players must be in the game.")
            raise ValueError("Both players must be in the game.")

        if card not in self.players[from_player]:
            logging.error(f"Card {card} not found in {from_player}'s hand.")
            raise ValueError(f"Card {card} not found in {from_player}'s hand.")

        # Transfer all cards from the from_player to the to_player.
        if card in self.players[from_player]:
            self.players[to_player].append(card)
            self.players[from_player].remove(card)
            logging.info(f"Card {card} transferred from {from_player} to {to_player}.")
            self.logs.append({"FORCE_SHIFT":f"Clean-up; shifted {card} from {from_player} to {to_player}."})

        else:
            raise ValueError(f"Card {card} not found in {from_player}'s hand.")
    
    # Reset the game state for a new game.
    def reset_game_state(self):
        self.players = {}
        self.teams = {}
        self.caller = None
        self.dropped_pits = []
        self.logs = []
        logging.info("Game state has been reset.")

